README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ROptionsStrategies

<!-- badges: start -->

[![Travis build
status](https://travis-ci.com/pohnson/ROptionsStrategies.svg?branch=main)](https://travis-ci.com/pohnson/ROptionsStrategies)
<!-- badges: end -->

The goal of ROptionsStrategies is to utilize the flexibilities of R and
its vast libraries to analyze options data provided through TDAmeritrade
API. Although the only available strategy is short put (The Wheel,)
strategies such as spread, iron condor, and butterfly are in the
development roadmap.

## Installation

You can install the released version of ROptionsStrategies from
[CRAN](https://CRAN.R-project.org) WHEN IT IS AVAILABLE with:

``` r
#install.packages("ROptionsStrategies") Not available yet
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("pohnson/ROptionsStrategies")
```

## Authentication

The most important data point in this package is the TDAmeritrade’s
Consumer Key. If you haven’t had one yet, please follow this thorough
guideline to obtain one:
<a href="https://www.reddit.com/r/algotrading/comments/c81vzq/td_ameritrade_api_access_2019_guide/">link</a>.

## Wheel\_Extractor Function

Now, let’s assume that you already have one and save it in a csv file.
We then can read it like this.

``` r
library(ROptionsStrategies)
library(tidyverse)

consumer_key <- read_csv("consumer_key.csv") %>% as.character()
```

Let’s assume that you can only execute one Wheel trade and wonder if the
option should be AMD or XLF. You can use `wheel_extractor` to get
underlying data as follows:

``` r
data <- wheel_extractor(
  ticker = c('AMD', 'XLF'),
  expiration_date = '2021-08-06',
  consumer_key = consumer_key,
  sleep_second = 1
)
#> [1] "Requesting data for AMD"
#> [1] "Requesting data for XLF"
#> [1] "Finished requesting data from TDAmeritrade."

head(data)
#>    bid  delta strikePrice inTheMoney ticker prob_otm cash_inflow
#> 1 0.00 -0.002          55      FALSE    AMD    0.998           0
#> 2 0.00 -0.003          60      FALSE    AMD    0.997           0
#> 3 0.01 -0.006          65      FALSE    AMD    0.994           1
#> 4 0.00 -0.006          66      FALSE    AMD    0.994           0
#> 5 0.00 -0.009          67      FALSE    AMD    0.991           0
#> 6 0.00 -0.007          68      FALSE    AMD    0.993           0
#>   invested_capital return_on_risk expiration_date
#> 1             5500   0.0000000000      2021-08-06
#> 2             6000   0.0000000000      2021-08-06
#> 3             6500   0.0001538462      2021-08-06
#> 4             6600   0.0000000000      2021-08-06
#> 5             6700   0.0000000000      2021-08-06
#> 6             6800   0.0000000000      2021-08-06
```

But you can also parse data from another data frame using `$`:

``` r
data_ticker <- data.frame(
  ticker = c('AMD', 'XLF')
)

data2 <- wheel_extractor(
  ticker = data_ticker$ticker,
  expiration_date = '2021-08-06',
  consumer_key = consumer_key,
  sleep_second = 1
)
#> [1] "Requesting data for AMD"
#> [1] "Requesting data for XLF"
#> [1] "Finished requesting data from TDAmeritrade."

head(data2)
#>    bid  delta strikePrice inTheMoney ticker prob_otm cash_inflow
#> 1 0.00 -0.002          55      FALSE    AMD    0.998           0
#> 2 0.00 -0.003          60      FALSE    AMD    0.997           0
#> 3 0.01 -0.006          65      FALSE    AMD    0.994           1
#> 4 0.00 -0.006          66      FALSE    AMD    0.994           0
#> 5 0.00 -0.009          67      FALSE    AMD    0.991           0
#> 6 0.00 -0.007          68      FALSE    AMD    0.993           0
#>   invested_capital return_on_risk expiration_date
#> 1             5500   0.0000000000      2021-08-06
#> 2             6000   0.0000000000      2021-08-06
#> 3             6500   0.0001538462      2021-08-06
#> 4             6600   0.0000000000      2021-08-06
#> 5             6700   0.0000000000      2021-08-06
#> 6             6800   0.0000000000      2021-08-06
```

## Wheel Visualizer

Then you can use `wheel_visualizer` to inspect which stock at a given
risk (Probability OTM) has the highest return.

``` r
wheel_visualizer(
  data = data,
  include_in_the_money = FALSE
)
```

<img src="man/figures/README-visualize-1.png" width="100%" />


NEWS.md
# ROptionsStrategies 1.1.0

## Release Notes and News

### ROptionsStrategies 1.1.0
Initial release includes basic functionality to help with the Wheel trading strategy. More advanced functionalities for the Wheel and other strategies such as Iron Condor or Spread are expected in subsequent releases.


R/wheel_extractor.R
#' Find the Best Options for The Wheel Strategy
#'
#' Create a data frame with options information to determine
#' what options is the best for short put also known as "The Wheel" strategy
#'
#' @param ticker stock ticker
#' @param expiration_date an option expiration date
#' @param consumer_key Consumer Key from TDAmeritrade Developer
#' @param sleep_second Sleep in second before making a new API call
#'
#' @return
#' A data frame object with 10 columns: bid, delta, strikePrice, inTheMoney, ticker, prob_otm,
#' cast_inflow, invested_capital, return_on_risk, and expiration_date. You can inspect the result by yourself or use wheel_visualizer().
#'
#' @examples
#' #You want to know what strike price from AAPL, AMZN, and TSLA whose
#' #expiration date is Dec 31, 2021 has the highest return on risk.
#' #You'd input the parameters as follows:
#'
#' #result <- wheel_extractor(
#' #      ticker = c('AAPL', 'AMZN', 'TSLA'),
#' #      expiration_date = '2021-12-31',
#' #      consumer_key = 'Your TDA API Key',
#' #      sleep_second = 1 #or 0 if you want
#' # )
#'
#' #If you have many stocks that you want to know you can simply create a data frame and use $.
#'
#' ###### First create a data frame #####
#' candidates <- data.frame(ticker = c('AAPL', 'AMZN', 'TSLA', 'GOOG'))
#'
#' ###### Then use $ #####
#' #result <- wheel_extractor(
#' #      ticker = candidates$ticker
#' #      expiration_date = '2021-12-31',
#' #      consumer_key = 'xxx',
#' #      sleep_second = 1 #or 0 if you want
#' # )
#'
#' @references
#' If you haven't created a free TDAmeritrde API Key, please follow steps from this website https://www.reddit.com/r/algotrading/comments/c81vzq/td_ameritrade_api_access_2019_guide/
#'
#' @seealso
#' Execute ?wheel_visualizer() on how to visualize the resulting data frame
#' @export
#' @importFrom magrittr %>%
#' @import httr
#' @importFrom rlang .data


wheel_extractor <- function(
  ticker = c('AAPL','GOOG'),
  expiration_date = Sys.Date(),
  consumer_key = NULL,
  sleep_second = 1
){

  #Check if ticker input is in a correct format
  if(class(ticker) != 'character'){
    stop('The input is not in a correct format. Please input either in vector or data frame with $ specifying a ticker column.')
  } # End of ticker check if


  #Check if date is more than 1
  if(length(expiration_date) > 1){
    stop('Please only input one expiration date')
  }

  #Check if date is in a correct format
  if(class(try(as.Date(expiration_date))) == 'try-error'){
    stop('Incorrect input expiration date format. Please input expiration date in YYYY-MM-DD format.')
  }

  #Check if consumer key is blank
  if(is.na(consumer_key) == TRUE || is.null(consumer_key) == TRUE){
    stop('Please input consumer key from TDAmertirade Developer.')
  }

  #Convert ticker to upper case
  ticker <- toupper(ticker)

  ##### Start the Loop to Extract Data #####
  temp_container <- data.frame(NULL)
  for(loop_a in 1:length(ticker)){
    url_1 <- 'https://api.tdameritrade.com/v1/marketdata/chains?apikey='
    url_api_key_1.1 <- consumer_key
    url_1.2 <- '&symbol='
    url_stock_2 <- ticker[loop_a]
    url_3 <- '&contractType=PUT&includeQuotes=TRUE&strategy=ANALYTICAL&range=ALL&fromDate='
    url_from_4 <- expiration_date
    url_5 <- '&toDate='
    url_to_6 <- expiration_date
    url_final <- paste(url_1, url_api_key_1.1, url_1.2, url_stock_2, url_3, url_from_4, url_5, url_to_6, sep = '')

    print(
      paste(
        'Requesting data for',
        url_stock_2,
        sep = ' '
      )
    )

    ##### Request data from TDA #####
    temp_data <- try(
      httr::GET(
        url = url_final
      )
    )

    ##### Check if the result has HTTP Code 200 #####
    if(temp_data[['status_code']] != 200){
      stop(
        'The error code is ', temp_data[['status_code']], '.\n ',
        'Error Code 400 indicates an error message indicating the caller must pass a non null value in the parameter.\n',
        'Error Code 401 indicates unauthorized.\n',
        'Error Code 403 indicates forbidden.\n',
        'Error Code 404 indicates optionchain for the symbol was not found.\n',
        'Error Code 500 indicates internal error in the service.'
      )
    } #End Check for Status Code

    ##### Extract Content #####
    temp_data_2 <- httr::content(temp_data)

    ##### Extracting Data from List #####
    temp_container_2 <- data.frame(NULL)
    if(temp_data_2[[2]] == 'SUCCESS'){

      temp_data_3 <- temp_data_2[c('putExpDateMap')]

      for(loop_b in 1:length(temp_data_3)){
        y <- list()

        for(loop_c in 1:length(temp_data_3[[loop_b]])){ #at date time level

          for(loop_d in 1:length(temp_data_3[[loop_b]][[loop_c]])){ #at strike price level
            y <- append(
              y,
              temp_data_3[[loop_b]][[loop_c]][[loop_d]] #data at strike level
            )

          } #End of loop_d

          ##### Get Only Desired Columns #####
          y_2 <- lapply(y, helper_list_to_select)

          y_3 <- data.table::rbindlist(y_2, fill = T) %>%
            dplyr::mutate(ticker = temp_data_2$symbol)

        } # End of loop_c

        #Combine Data from Each Expiration of a Given Stock
        temp_container_2 <- rbind(temp_container_2, y_3)

      }# End of loop_b

    }# End If Success

    #Combine Data from Different Stocks
    temp_container <- rbind(temp_container, temp_container_2)

    #Sleep to make sure that TDAmeritrade will not disconnect due to too many requests
    Sys.sleep(sleep_second)

  } #End of loop_a
  print('Finished requesting data from TDAmeritrade.')

  temp_container_3 <- temp_container %>%
    dplyr::mutate(
      prob_otm = 1 + .data$delta,
      cash_inflow = .data$bid * 100,
      invested_capital = .data$strikePrice*100,
      return_on_risk = (.data$bid*100)/.data$invested_capital,
      expiration_date = expiration_date
    )

  return(data.frame(temp_container_3))

} # End of function

helper_list_to_select <- function(x){
  return(
    x[c('bid', 'delta', 'strikePrice', 'inTheMoney')]
  )
}



R/wheel_visualizer.R
#' Visualize Options with the Best Return on Risk
#'
#' This function will create a histogram plot for the resulting data frame from
#' the wheel_extractor() function.
#'
#' @param data A resulting data frame object from wheel_extractor() function.
#' @param visualization_type Currently only support ggplot2's ggplot_histogram.
#' @param include_in_the_money The Wheel generally employs on OTM options. But the data frame also includes ITM.
#'
#' @return
#' Ggplot2 histogram chart.
#'
#' @examples
#' #First let's create sample data frame
#' sample <- data.frame(
#'      strikePrice = c(50, 60, 70, 80),
#'      inTheMoney = c(FALSE, FALSE, FALSE, TRUE),
#'      ticker = c('AAPL', 'AAPL', 'TSLA', 'TSLA'),
#'      prob_otm = c(0.3, 0.4, 0.3, 0.4),
#'      cash_inflow = c(250, 400, 115, 220),
#'      invested_capital = c(5000, 6000, 7000, 8000),
#'      return_on_risk = c(0.05, 0.07, 0.02, 0.03)
#' )
#'
#' #Then use the function
#' wheel_visualizer(data = sample)
#'
#' @export
#' @import ggplot2
#' @importFrom magrittr %>%
#' @importFrom scales percent
#' @importFrom rlang .data

wheel_visualizer <- function(
data = NULL,
visualization_type = 'ggplot_histogram',
include_in_the_money = FALSE
){
  ##### Check if input data frame has necessary columns #####
  if(
    sum(
      c(
        'strikePrice', 'inTheMoney', 'ticker', 'prob_otm', 'cash_inflow',
        'invested_capital', 'return_on_risk'
      ) %in% names(data)
    ) < 7
  ){
    stop('The input data frame does not have required columns: strikePrice, inTheMoney, ticker, prob_otm, cash_inflow, invested_capital, and return_on_risk')
  } # End Check Column

  ##### Check if include_in_the_money is either T or F #####
  if(is.logical(include_in_the_money) == FALSE){
    stop('Include_In_The_Money value can only be either TRUE or FALSE.')
  }

  ##### Check if Visualization Type is Valid #####
  if(visualization_type != 'ggplot_histogram'){
    stop('Current function can only visualize histogram using GGPLOT2 function.')
  }

  ##### ITM Filtering #####
  if(include_in_the_money == FALSE){
    data_final <- data %>% dplyr::filter(.data$inTheMoney == FALSE)
  } else {
    data_final <- data
  }

  ##### Visualize #####
  if(visualization_type == 'ggplot_histogram'){
    data_final %>%
      dplyr::filter(
        .data$prob_otm > 0
      ) %>%
      ggplot2::ggplot(
        ggplot2::aes(y = .data$prob_otm, x = .data$return_on_risk, col = .data$ticker, label = .data$strikePrice)
      ) +
      geom_point() +
      ggplot2::scale_y_continuous(labels = scales::percent) +
      ggplot2::scale_x_continuous(labels = scales::percent) +
      ggplot2::xlab('Return on Risk') +
      ggplot2::ylab('Probablity OTM')
      #ggplot2::geom_text()
      #ggrepel::geom_label_repel(max.overlaps = 200)

  } #End If visualization_type

} #End wheel_visualizer function


